backend/routes/session.ts

```diff
@@ -1,38 +1,46 @@
 import { Request, Response, Router } from 'express';
-import createVideoService from '../videoServiceFactory';
+import createVideoService from '../videoService/videoServiceFactory';
 import getSessionStorageService from '../sessionStorageService';
+import blockCallsForArgs from '../helpers/blockCallsForArgs';
 
 const sessionRouter = Router();
 const videoService = createVideoService();
 const sessionService = getSessionStorageService();
 
-sessionRouter.get('/:room', async (req: Request, res: Response) => {
+/**
+ * Gets or creates a session in a thread safe (async operation safe) way.
+ * The function is wrapped in blockCallsForArgs to ensure that simultaneous calls for a room name return the same session.
+ * Not blocking these calls would result in separate sessions being created for the same room name.
+ * @param {string} roomName - name of the meeting room
+ * @returns {Promise<string>} the sessionId
+ */
+export const getOrCreateSession = blockCallsForArgs(async (roomName: string) => {
+  let sessionId = await sessionService.getSession(roomName);
+  if (!sessionId) {
+    const session = await videoService.createSession();
+    sessionId = session.sessionId;
+    await sessionService.setSession(roomName, sessionId);
+  }
+  return sessionId;
+});
+
+sessionRouter.get('/:room', async (req: Request<{ room: string }>, res: Response) => {
   try {
     const { room: roomName } = req.params;
-    const sessionId = await sessionService.getSession(roomName);
-    if (sessionId) {
-      const data = videoService.generateToken(sessionId);
-      res.json({
-        sessionId,
-        token: data.token,
-        apiKey: data.apiKey,
-      });
-    } else {
-      const data = await videoService.getCredentials();
-      await sessionService.setSession(roomName, data.sessionId);
-      res.json({
-        sessionId: data.sessionId,
-        token: data.token,
-        apiKey: data.apiKey,
-      });
-    }
+    const sessionId = await getOrCreateSession(roomName);
+    const data = videoService.generateToken(sessionId);
+    res.json({
+      sessionId,
+      token: data.token,
+      apiKey: data.apiKey,
+    });
```

Comment:
I refactored this to:
- Split phase - separate session creation from token creation
- wrap getOrCreate session in the async blocker wrapper